File: perf.c
#include <time.h>
#include <stdio.h>
#include <stdlib.h>

double clock_now()
{
  struct timespec now;
  clock_gettime(CLOCK_MONOTONIC, &now);
  return (double)(now.tv_sec) + (double)(now.tv_nsec/1.0e9);
}

int fib(int n) {
  return n < 2 ? n : fib(n-1) + fib(n-2);
}

int iterative_fib(int n) {
  if (n < 2) return 1;
  
  int sum = 0, i;

  for(i = 1; i <= n; i++) {
    sum += i;
  } 

  return sum;
}

double *rand_double_array(int n) {
  double *d = (double *)malloc(n*sizeof(double));
  for(int i = 0; i <n; i++) {
    d[i] = rand()%1;
  }
  return d;
}

void quicksort(double *a, int lo, int hi) {
  int i = lo;
  int j = hi;
  while (i < hi) {
    double pivot = a[(lo+hi)/2];
    // Partition
    while (i <= j) {
      while (a[i] < pivot) {
        i = i + 1;
      }
      while (a[j] > pivot) {
        j = j - 1;
      }
      if (i <= j) {
        double t = a[i];
        a[i] = a[j];
        a[j] = t;
        i = i + 1;
        j = j - 1;
      }
    }

    // Recursion for quicksort
    if (lo < j) {
      quicksort(a, lo, j);
    }
    lo = i;
    j = hi;
  }
}

double pisum() {
  double sum = 0.0;
  for (int j=0; j<500; ++j) {
    sum = 0.0;
    for (int k=1; k<=10000; ++k) {
      sum += 1.0/(k*k);
    }
  }
  return sum;
}

void printfd(int n) {
  FILE *f = fopen("/dev/null", "w");
  long i = 0;
  for (i = 0; i < n; i++)
    fprintf(f, "%ld %ld\n", i, i+1);
  fclose(f);
}

long parse_int(const char *s, long base) {
  long n = 0;
  for (; *s; ++s) {
    char c = *s;
    long d = 0;
    if (c >= '0' && c <= '9') d = c-'0';
    else if (c >= 'A' && c <= 'Z') d = c-'A' + (int) 10;
    else if (c >= 'a' && c <= 'z') d = c-'a' + (int) 10;
    else exit(-1);

    if (base <= d) exit(-1);
    n = n*base + d;
  }
  return n;
}

void print_perf(char *test, double time) {
  printf("%s,%s,%.9f\n", "C", test, time);
}

int main(int argc, char **argv) { 
  double t;

  tmin = INFINITY;
  for (int i=0; i<NITER; ++i) {
    t = clock_now();
    fib(30);
    t = clock_now()-t;
    if (t < tmin) tmin = t;
  }
  print_perf("recursion_fibonacci", t);

  tmin = INFINITY;
  for (int i=0; i<NITER; ++i) {
    t = clock_now();
    iterative_fib(30);
    t = clock_now()-t;
    if (t < tmin) tmin = t;
  }
  print_perf("iteration_fibonacci", t);

  tmin = INFINITY;
  for (int i=0; i<NITER; ++i) {
    t = clock_now();
    pisum();
    t = clock_now()-t;
    if (t < tmin) tmin = t;
  }
  print_perf("iteration_pi_sum", t);

  tmin = INFINITY;
  for (int i=0; i<NITER; ++i) {
    t = clock_now();
    printfd(100000);
    t = clock_now()-t;
    if (t < tmin) tmin = t;
  }
  print_perf("print_to_file", t);

  tmin = INFINITY;
  for (int i=0; i<NITER; ++i) {
    double *d = rand_double_array(5000);
    t = clock_now();
    quicksort(d, 0, 5000-1);
    t = clock_now()-t;
    if (t < tmin) tmin = t;
  }
  print_perf("imperative_recursion_quicksort", t);

  tmin = INFINITY;
  for (int i=0; i<NITER; ++i) {
    t = clock_now();
    char s[11];
    for (int k=0; k<1000; ++k) {
      uint32_t n = rand();
      sprintf(s, "%x", n);
      uint32_t m = (uint32_t)parse_int(s, 16);
      assert(m == n);
    }
    t = clock_now()-t;
    if (t < tmin) tmin = t;
  }
  print_perf("parse_integers", t);
  
  return 0;
}

File: perf.jl
# This file was formerly a part of Julia. License is MIT: https://julialang.org/license
# https://github.com/JuliaLang/Microbenchmarks/blob/master/perf.jl

# using Base
using Printf
using Random
using BenchmarkTools

import LinearAlgebra

macro print_elapsed(f, name, desc)
    name = esc(name)
    desc = esc(desc)
    quote 
        Printf.@printf "julia,%s,%.9f\n" $name (@belapsed $f)
    end
end

## recursive fib ##
fib(n) = n < 2 ? n : fib(n-1) + fib(n-2)

@print_elapsed fib(20) "recursion_fibonacci" "Recursive fibonacci"


## tail recursive fib ##
helper(current, next, n) = n==0 ? current : helper(next, next+current, n-1)
tr_fib(n) = helper(Int(0), Int(1), n)

@print_elapsed tr_fib(20) "tail_recursive_fibonacci" "Recursive fibonacci"


## parse integer ##
# function parseintperf(t)
#     local n, m
#     for i=1:t
#         n = rand(UInt32)
#         s = string(n, base = 16)
#         m = UInt32(parse(Int64, s, base = 16))
#         @assert m == n
#     end
#     return n
# end

# @print_elapsed parseintperf(1000) "parse_integers" "Integer parsing"


## mandelbrot set: complex arithmetic and comprehensions ##
# function myabs2(z)
#     return real(z)*real(z) + imag(z)*imag(z)
# end

# function mandel(z)
#     c = z
#     maxiter = 80
#     for n = 1:maxiter
#         if myabs2(z) > 4
#             return n-1
#         end
#         z = z^2 + c
#     end
#     return maxiter
# end

# mandelperf() = [ mandel(complex(r,i)) for i=-1.:.1:1., r=-2.0:.1:0.5 ]
# @print_elapsed mandelperf() "userfunc_mandelbrot" "Calculation of mandelbrot set"


## numeric vector sort ##
function qsort!(a,lo,hi)
    i, j = lo, hi
    while i < hi
        pivot = a[(lo+hi)>>>1]
        while i <= j
            while a[i] < pivot; i += 1; end
            while a[j] > pivot; j -= 1; end
            if i <= j
                a[i], a[j] = a[j], a[i]
                i, j = i+1, j-1
            end
        end
        if lo < j; qsort!(a,lo,j); end
        lo, j = i, hi
    end
    return a
end

sortperf(n) = qsort!(rand(n), 1, n)
@print_elapsed sortperf(5000) "imperative_recursion_quicksort" "Sorting of random numbers using quicksort"


## slow pi series ##
function pisum()
    sum = 0.0
    for j = 1:500
        sum = 0.0
        for k = 1:10000
            sum += 1.0/(k*k)
        end
    end
    sum
end

@print_elapsed pisum() "iteration_pi_sum" "Summation of a power series"


## random matrix statistics ##
# function randmatstat(t)
#     n = 5
#     v = zeros(t)
#     w = zeros(t)
#     for i=1:t
#         a = randn(n,n)
#         b = randn(n,n)
#         c = randn(n,n)
#         d = randn(n,n)
#         P = [a b c d]
#         Q = [a b; c d]
#         @static if VERSION >= v"0.7.0" 
#             v[i] = LinearAlgebra.tr((P'*P)^4)
#             w[i] = LinearAlgebra.tr((Q'*Q)^4)
#         else
#             v[i] = trace((P'*P)^4)
#             w[i] = trace((Q'*Q)^4)
#         end
#     end
# end

# @print_elapsed randmatstat(1000) "matrix_statistics" "Statistics on a random matrix"


# ## largish random number gen & matmul ##
# @print_elapsed rand(1000,1000)*rand(1000,1000) "matrix_multiply" "Multiplication of random matrices"


## printfd ##
function printfd(n)
    open("/dev/null", "w") do io
        for i = 1:n
            Printf.@printf(io, "%d %d\n", i, i + 1)
        end
    end
end

@print_elapsed printfd(100000) "print_to_file" "Printing to a file descriptor"
